### DISCUSSION TITLE
How do you usually restore a deleted file?

### DISCUSSION THREAD

**Discussion by iamFIREcracker** [tpope/vim-fugitive#2166] [Q&A]
Here is what I did, today, but was curious to see if there were a better way of doing it.

1. Find the commit which deleted the file (either via `:Gclog ...`, or with `:G`), and open it (`<CR>`, `gO`, `O`, you name it)
3. Copy the commit id with `y<C-G>`
4. Find the chunck where the file got deleted, and then open the deleted file by pressing `<CR>` on its path
5. `:Ge` to edit the same file in work tree
6. `:Gread` incantation (i.e., `:Gread <C-R>0^:%`) to finally load the content of the deleted file

    - `<C-R>0` loads the commit ID from step 3
    - `^` is to select the ancestor
    - `:%` is to load the content of the file (without it, it would dump the commit message into the buffer)

Now, I can live with this: it's not like I am spending my day restoring files; but I was wondering if there wasn't a simpler way of doing this; maybe something that did not require step 2?

**Comment by iamFIREcracker** [comment]
Turns out one could use the `!` fugitive object for this!

From the help `fugitive-object`'s help:

```
!               The commit owning the current file
!:Makefile      The file named Makefile in the commit owning the current file
```

I guess I can get rid of step 2, and change 5 to: `:Gread !^:%`.

**Reply by tpope** [reply]
If you're in a work-tree file, `!` is just an alias for `@` aka `HEAD`. So you can also do `:Gread @^:%`.

**Comment by tpope** [comment]
Alternatively, if you're already editing the file you want to restore in the history (e.g., via `:Gedit @^:%`), you can call `:Gwrite!` and it will write the contents of the current buffer to the work-tree and the index. This method makes more sense if you're already rooting around in the history looking for where to restore the file from.
